### ISSUE TITLE
rest.cors.allowed-origins is set to narrow if the UI uses any namespace (subdirectory)

### ISSUE THREAD

**Issue by pnbecker** [DSpace/DSpace#9131]
**TLDR;**
When the ui runs in a path and not directly at the root of an URL, `rest.cors.allowed-origins` will be set by default to the `dspace.ui.url` which will inlcude the path of the URL. Browsers do not include the path in their ORIGIN headers and the request will fail with a CORS error. As a workaround you must have to set `rest.cors.allowed-origins` to the ui url containing the protocol, hostname and port but nothing further.

**Describe the bug**
Imagine you run DSpace's ui at `https://example.org:1234/dspace`. You will set `dspace.ui.url = https://example.org:1234/dspace` so that links generated by the backend are pointing towards your UI. By default `rest.cors.allowed-origins` is set to `dspace.ui.url` and therefore will be set to `https://example.org:1234/dspace`. But Browsers do not include paths in their origin header:

> An \<origin\> is a combination of scheme, host and port. Unlike HTTP Referer, no path data or query string will be provided in the origin.
> https://wiki.mozilla.org/Security/Origin#Origin_header_format

Requests will end in a http 500 error and you will see in your browsers dev tool, that requests to the backend fail in cause of a cors.

**To Reproduce**
1. Setup a DSpace instance running the ui in a subdirectory, using a ui namespace.
2. Set dspace.ui.url to the url of your user interface, including the path (see the example above).
3. Open up a browser, open up the network section of the dev tools, open DSpace's UI in the browser and what out for requests towards the backend. See that they contain `CORS Missing Allow Origin` or `NS_ERROR_DOM_BAD_URI` (or something comparable depending on your browser). Check the Origin header that was send with your request.
4. In the backend run `/dspace/bin/dspace dsprop -p rest.cors.allowed-origins` to check that `rest.cors.allowed-origins` contains the whole ui url, including the ui namespace/path.
5. Set `rest.cors.allowed-origins` in your backend to `protocol://url:port` of the ui (for example `http://localhost:4000/dspace`) and restart the backend.
6. Reload the ui in your browser.

**Expected behavior**
We either automatically strip any paths that come up in `rest.cors.allowed-origins` or we document better, that you have to change this configuration if you install DSpace not at the root-level of an URL. We also could break up dspace.ui.url into multiple variables as it is done in the frontend configuration.

**Related work**
https://github.com/DSpace/DSpace/pull/3171

**Comment by pnbecker** [comment]
@tdonohue as you worked on #3171 I would be very interested in your thoughts about this.

**Comment by pnbecker** [comment]
https://wiki.lyrasis.org/display/DSDOC7x/REST+API states:
> Keep in mind any URLs added to this setting must be an exact match with the origin: mode (http vs https), domain, port, and subpath(s) all must match.. So, for example, these URLs are all considered different origins: "http://mydspace.edu", "http://mydspace.edu:4000" (different port), "https://mydspace.edu" (http vs https), "https://myapp.mydspace.edu" (different domain), and "https://mydspace.edu/myapp" (different subpath).

In my experience that is not correct. Browsers doesn't add the path to the header `Origin`. See also my quote from mozilla above in the issue description.

**Comment by tdonohue** [comment]
@pnbecker : My immediate feedback is that the **behavior of** `rest.cors.allowed-origins` _is not under DSpace's control_.  It's  just a way to set `allowedOrigins()` in Spring Security / Spring Boot here: https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/main/java/org/dspace/app/rest/Application.java#L195

At the time I wrote [those docs](https://wiki.lyrasis.org/display/DSDOC7x/REST+API), I wrote them based on what I was able to figure out about Spring's `allowedOrigins()` behavior. That said, it is possible I that the documentation I wrote was incorrect in some manner.  That'd require further testing/verification from someone.

It obviously would possible for us to strip subpaths, if we find that provides better "default" behavior.  I've not tried that myself, but I do know that we had to write some custom code to strip trailing slashes, as it was found that a trailing slash caused issues in Spring Security's own Origin checks. See the code added here: https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/ApplicationConfig.java#L83-L90

Overall, I don't have anything else to add.  Just wanted to note that Spring Security controls much of the code logic here.  So, to solve this issue might require someone digging deeper into whether our documentation is misleading and/or whether it would be better to strip subpaths before sending them to Spring Security.

**Comment by OyvindLGjesdal** [comment]
I noticed that this still applies, and found this issue when googlingen.

Changed backend url from  `https://example-frontend.org`  (which worked) to `https://example-frontend.org/repository`  which caused a CORS error.

I see there already is a note in the docs

```
NOTE #3: Although subpath must match, the Origin header itself sent from Angular will never contain a subpath. So if the dspace.ui.url  config property ever changes to include a subpath like /myapp, then the expected origin will need to be added to rest.cors.allowed-origins  , ie. the URL without  the subpath.
```

This works when setting the rest.cors-allowed-origins,so maybe it is a non-issue, but stripping the subpath from the uiURL, when there aren't any rest.cors.allowed-origins , would be a reasonable default/fallback? Something like, I think the logic would hold:

https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/ApplicationConfig.java#L83-L90
```
# replace everything after third / includin, if matches 
uiURL.replaceAll("([^/]+//[^/]+)/.+","$1")
```
